File: app/src/main/java/com/example/ukeje/countrypedia/adapters/SearchResultListAdapter.java
package com.example.ukeje.countrypedia.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.ukeje.countrypedia.R;
import com.example.ukeje.countrypedia.web.responses.CountryResponse;

import java.util.List;

/**
 * @author .: Ukeje Emeka
 * @email ..: ukejee3@gmail.com
 * @created : 6/26/19
 */
public class SearchResultListAdapter extends RecyclerView.Adapter<SearchResultListAdapter.MyViewHolder> {

    Context mContext;
    List<CountryResponse> list;
    View view;
    CountryListClickListener listener;

    public SearchResultListAdapter(Context context, List<CountryResponse> list, CountryListClickListener listener){

        this.mContext = context;
        this.list = list;
        this.listener = listener;
    }

    public List<CountryResponse> getList(){
        return list;
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_search_result_list, parent, false);
        MyViewHolder myViewHolder = new MyViewHolder(view);
        return myViewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {

        CountryResponse country = list.get(position);
        holder.countryName.setText(country.getName());
    }

    public void refreshList(List<CountryResponse> newList){
        list = newList;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    class MyViewHolder extends RecyclerView.ViewHolder{

        TextView countryName;

        public MyViewHolder(@NonNull View itemView) {
            super(itemView);
            countryName = itemView.findViewById(R.id.result_country_name);

            itemView.setOnClickListener(v -> {
                if(listener != null)
                    listener.onItemClick(list.get(getAdapterPosition()));
            });
        }
    }

    public interface CountryListClickListener{
        void onItemClick(CountryResponse countryResponse);
    }
}

File: app/src/main/java/com/example/ukeje/countrypedia/adapters/SearchResultListAdapter.kt
package com.example.ukeje.countrypedia.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.ukeje.countrypedia.R
import com.example.ukeje.countrypedia.web.responses.CountryResponse

/**
 * @author .: Ukeje Emeka
 * @email ..: ukejee3@gmail.com
 * @created : 6/26/19
 */
class SearchResultListAdapter(var list: List<CountryResponse>, private var listener: (countryResponse: CountryResponse) -> Unit) : RecyclerView.Adapter<SearchResultListAdapter.MyViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {

       val view = LayoutInflater.from(parent.context).inflate(R.layout.item_search_result_list, parent, false)
        return MyViewHolder(view)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {

        val (name) = list[position]
        holder.countryName.text = name
    }

    fun refreshList(newList: List<CountryResponse>) {
        list = newList
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return list.size
    }

    inner class MyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var countryName: TextView = itemView.findViewById(R.id.result_country_name)

        init {

            itemView.setOnClickListener {
                listener(list[adapterPosition])
            }
        }
    }
}
